### PR TITLE
Diag

### LINKED ISSUE #17024: [Operator] add diag op
test=develop
Like numpy.diag, this op generates a square matrix with specified diagonal values.

### PR DESCRIPTION
**Pull request by zhoukunsheng**
fix #17024

### CODE REVIEW

**Comment by zhoukunsheng** [comment]
image

**Review by shanyi15 — DISMISSED** [review]
LG comments

File: paddle/fluid/operators/diag_op.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline paddle/fluid/operators/diag_op.cc]: 2016 -> 2019

File: paddle/fluid/operators/diag_op.cc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                   "Input(Diagonal) of DiagOp should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("Out"),
+                   "Output(Out) of DiagOp should not be null.");
+
+    auto s_dims = ctx->GetInputDim("Diagonal");
+    PADDLE_ENFORCE(s_dims.size() == 1,
+                   "The rank of Input(Diagonal) should only be 1.");
+
+    ctx->SetOutputDim("Out", {s_dims[0], s_dims[0]});
```
Comment (chengduoZH) [inline paddle/fluid/operators/diag_op.cc:35]: Please set the LoD for `Out`.

File: paddle/fluid/operators/diag_op.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* diagonal = context.Input<framework::Tensor>("Diagonal");
+    auto* diag_data = diagonal->data<T>();
+    auto numel = diagonal->numel();
+    auto* out = context.Output<framework::Tensor>("Out");
+    T* out_data = out->mutable_data<T>(context.GetPlace());
+
+#ifdef __NVCC__
+    thrust::fill(thrust::device_pointer_cast(out_data),
+                 thrust::device_pointer_cast(out_data) + numel * numel, 0);
```
Comment (chengduoZH) [inline paddle/fluid/operators/diag_op.h]: Please use
```
    math::SetConstant<DeviceContext, T> set_constant;
   set_zero(dev_ctx, input_grad, static_cast<T>(0));
```

File: python/paddle/fluid/layers/tensor.py [hunk 4]
```diff
@@ -884,3 +884,41 @@ def zeros_like(x, out=None):
         type='fill_zeros_like', inputs={'X': [x]}, outputs={'Out': [out]})
     out.stop_gradient = True
     return out
+
+
+def diag(diagonal, out=None):
```
Comment (chengduoZH) [inline python/paddle/fluid/layers/tensor.py]: Don't make `out` as the parameter here.

File: paddle/fluid/operators/diag_op.cc (hunk 1 above)
Comment (zhoukunsheng) [reply paddle/fluid/operators/diag_op.cc]: Sure.

File: paddle/fluid/operators/diag_op.cc (hunk 2 above)
Comment (zhoukunsheng) [reply paddle/fluid/operators/diag_op.cc:35]: @chengduoZH right now this op doesn't support LoD. Do you think we need to support it ?

File: paddle/fluid/operators/diag_op.h (hunk 3 above)
Comment (zhoukunsheng) [reply paddle/fluid/operators/diag_op.h]: sure, thanks.

File: python/paddle/fluid/layers/tensor.py (hunk 4 above)
Comment (zhoukunsheng) [reply python/paddle/fluid/layers/tensor.py]: ok, will do.

File: paddle/fluid/operators/diag_op.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#ifdef __NVCC__
+#include <thrust/fill.h>
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/operators/diag_op.h]: Please remove the unused code.

File: python/paddle/fluid/tests/unittests/test_diag.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_diag.py]: 2018->2019

File: python/paddle/fluid/layers/tensor.py [hunk 7]
```diff
@@ -884,3 +884,39 @@ def zeros_like(x, out=None):
         type='fill_zeros_like', inputs={'X': [x]}, outputs={'Out': [out]})
     out.stop_gradient = True
     return out
+
+
+def diag(diagonal):
```
Comment (sneaxiy) [inline python/paddle/fluid/layers/tensor.py:889]: Change `diagonal` to `x` or `input`?

File: paddle/fluid/operators/diag_op.h [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#ifdef __NVCC__
+#include <thrust/fill.h>
```
Comment (sneaxiy) [inline paddle/fluid/operators/diag_op.h]: Remove this header.

Comment (zhoukunsheng) [reply paddle/fluid/operators/diag_op.h]: ok, thanks!

File: paddle/fluid/operators/diag_op.h (hunk 5 above)
Comment (zhoukunsheng) [reply paddle/fluid/operators/diag_op.h]: done, thanks!

File: python/paddle/fluid/tests/unittests/test_diag.py (hunk 6 above)
Comment (zhoukunsheng) [reply python/paddle/fluid/tests/unittests/test_diag.py]: done, thanks!

File: python/paddle/fluid/layers/tensor.py (hunk 7 above)
Comment (zhoukunsheng) [reply python/paddle/fluid/layers/tensor.py:889]: x or input is good, but `diagonal` is more specific about the meaning of this input, so I think we may just leave it as it is.

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Review by phlrain — APPROVED** [review]
LGTM

**Review by chengduoZH — APPROVED** [review]

**Review by shanyi15 — APPROVED** [review]

**Pull request merged**